### ISSUE TITLE
`rw test api` Jest memory performance and benchmarks; api test leaking

### ISSUE THREAD

**Issue by thedavidprice** [redwoodjs/redwood#4360]
### Background
Originally reported by @zygopleural here:
- #4208 

Web memory leak resolved and API memory leak mitigated via:
- #4096 

Because the API memory leak is not fully resolved, we want to benchmark and watch performance as well as keep this issue open for further work toward complete resolution.

This Issue uses the original reproduction with the new benchmarks.

## Expected behaviour

`yarn redwood test` should use ~ same heap size for each test unless the user has specifically done something wrong.

i.e. the heap size should **not** increase with each consecutive test or run (i.e. leak)

## What currently happens

`rw test web`:
The heap size is stable
> Note: /web not tested in the reproduction and example below. But [stability demonstrated here](https://github.com/redwoodjs/redwood/issues/4208#issuecomment-1028390697) after config updates

`rw test api`:
Even after mitigation in #4096 the heap size increases along with quantity of tests

### Option 1: Clone the Reproduction Repo
> Note: if cloned, do check the version of Redwood packages, which might need upgrading for testing
- https://github.com/thedavidprice/rw-jest-api-benchmark

#### 1. Creat App
`yarn create redwood-app --typescript test`

#### 2. Add some models to `./api/db/schema.prisma`
Excuse the silly models, just something quick to generate tests.

```prisma
model Man {
  id   String @id @default(cuid())
  name String
}

model Woman {
  id   String @id @default(cuid())
  name String
}

model Boy {
  id   String @id @default(cuid())
  name String
}

model Girl {
  id   String @id @default(cuid())
  name String
}

model Dog {
  id   String @id @default(cuid())
  name String
}

model Cat {
  id   String @id @default(cuid())
  name String
}
```

#### 3. Generate some tests
```shell
yarn redwood generate sdl --crud man
yarn redwood generate sdl --crud woman
yarn redwood generate sdl --crud boy
yarn redwood generate sdl --crud girl
yarn redwood generate sdl --crud dog
yarn redwood generate sdl --crud cat
```

#### 4. Run tests
`yarn redwood test --logHeapUsage`

### Initial test (watch mode)
<img width="578" alt="Screen Shot 2022-02-02 at 5 19 38 PM" src="https://user-images.githubusercontent.com/2951/152276915-dbfa2fc4-9303-4ddc-aef7-60e6ec6673c7.png">

### Second Run (watch
<img width="578" alt="Screen Shot 2022-02-02 at 5 20 09 PM" src="https://user-images.githubusercontent.com/2951/152276978-c0f05978-d46a-480a-928a-420b4227a069.png">
 mode)

### Third Run (watch mode)
<img width="578" alt="Screen Shot 2022-02-02 at 5 20 24 PM" src="https://user-images.githubusercontent.com/2951/152276988-2417c9b8-f83b-427f-a51d-be1779e0c8a6.png">

**Comment by callingmedic911** [comment]
**re: memory leak in web side tests**
#4208 had the changes to run web tests in parallel which reduced the frequency of leaks by running tests in parallel but in Github Action, where CPU is limited, the parallel execution is not always possible, which results in frequent crashes due to out of memory.
Here's an example repo: https://github.com/callingmedic911/rw-web-memory-leak with [CI check](https://github.com/callingmedic911/rw-web-memory-leak/pull/2) which fails on the latest Node LTS (16.14). Heap usage is better with Node 16.10 but the leak is still visible and if we increase the number of tests + mocks it'll also crash with Node 16.10 or lower.

Example CI run: https://github.com/callingmedic911/rw-web-memory-leak/runs/5321129020

This leak is exacerbated due to https://github.com/facebook/jest/issues/11956 but since we're require-ing all of mock in the project for each unit test across all test files, it's causing massive heap usage: https://github.com/redwoodjs/redwood/blob/59d31b3fca3fd6a247722d0b54f58a61b87e4c75/packages/testing/config/jest/web/jest.setup.js#L21-L25

I think for the next steps, we can evaluate:
1. (Context:  https://github.com/redwoodjs/redwood/pull/4457#issuecomment-1049990808)
instead of importing all mocks for each test, if we can somehow require only required ones. The idea is to put the import statement for a mock file inside the corresponding cell, so even if we have nested cells, the corresponding mock file would be imported. A good place to validate this would be by modifying the "auto-mock" babel plugin to insert the import statement **in cell**. 
2. as last resort we can `expose-gc` and force gc after each test suite run.

**Comment by callingmedic911** [comment]
Just realised I was supposed to put details in https://github.com/redwoodjs/redwood/issues/4356. This issue is related to api. Moving my comment to a separate issue.

**Comment by pi0neerpat** [comment]
Brb downloading more RAM
